README.md

```diff
@@ -1,134 +1,117 @@
-
-Another pattern that is beginning to surface is exporting async function and awaiting the results of imports, which drastically impacts our ability to do static analysis of the module graph.
+This pattern is appropriate for situations where loading a module is intended to schedule work that will happen some time later. However, the exports from this module may be accessed before this async function completes: If another module imports this one, it may see `output` as `undefined`, or it may see it after it's initialized to the return value of `process`, depending on when the access occurs! For example:
 
 ```mjs
-export default async () => {
-  // import other modules like this one
-  const import1 = await (await import('./import1.mjs')).default();
-  const import2 = await (await import('./import2.mjs')).default();
-
-  // compute some exports...
-  return {
-    export1: ...,
-    export2: ...
-  };
-};
-```
-
-At the current time [a search](https://github.com/search?utf8=%E2%9C%93&q=%22export+async+function%22&type=Code) for "export async function" on github produces over 5000 unique code examples of exporting an async function.
+// usage.mjs
+import { output } from "./awaiting.mjs";
+export function outputPlusValue(value) { return output + value }
 
-## Proposed solutions
-
-### Variant A: top-level `await` blocks tree execution
+console.log(outputPlusValue(100));
+setTimeout(() => console.log(outputPlusValue(100), 1000)
+```
 
-In this proposed solution a call to `top-level await` would block execution in the graph until it had resolved.
+### Workaround: Export a Promise to represent initialization
 
-In this implementation you could consider the following
+In the absense of this feature, it's possible to export a Promise from a module, and wait on that to know when its exports are ready. For example, the above module could be written as:
 
 ```mjs
-import a from './a.mjs'
-import b from './b.mjs'
-import c from './c.mjs'
-
-console.log(a, b, c);
+// awaiting.mjs
+import { process } from "./some-module.mjs";
+let output;
+export default (async () => {
+  const dynamic = await import(computedModuleSpecifier);
+  const data = await fetch(url);
+  output = process(dynamic.default, data);
+})();
+export output;
 ```
 
-If each of the modules above had a top level await present the loading would have similar execution order to
+Then, the module could be used as:
 
 ```mjs
-(async () => {
-  const a = await import('./a.mjs');
-  const b = await import('./b.mjs');
-  const c = await import('./c.mjs');
-  
-  console.log(a, b, c);
-})();
+// usage.mjs
+import default, { output } from "./awaiting.mjs";
+export function outputPlusValue(value) { return output + value }
+
+default.then(() => {
+  console.log(outputPlusValue(100));
+  setTimeout(() => console.log(outputPlusValue(100), 1000)
+});
 ```
 
-Module a would need to finish executing before b or c could execute.
+However, this leaves us with a number of problems still:
```

Comment:
Another problem: you need to separate declaration from initialisation. One could not write `export const example = await …`, but rather has to declare a top-level `export let/*sic!*/ example` and then initialise it in the IIAFE.

So while this would indeed be a proper workaround, nobody would use it. People, who already can often not distinguish destructurable objects from aliased import/export bindings, would instead default-export a promise that fulfills with a "module" object, and use
```js
import x from "x";
export default (async () => {
    const {xa, xb} = await x; // the parts of the import
    return { // the module "exports"
       exportA() { … }
       exportB: …
    };
})();
```

Reply:
Yeah, I agree that that is another unfortunate pattern that could emerge. Both patterns are bad; we could document both, but I thought this one would make the explainer clearer as it's closer to the solution.

Do we have any evidence that either of these two patterns is catching on? The GitHub search evidence that was previously used in the explainer seems rather weak.

Reply:
You can see the switch from `let` to `const` and the  combination of declaration and definition in the code sample I included. Would it really help to spell this out with further text?

Reply:
I have no idea how to use Github search effectively for design patterns. This approach is however what [I myself have](https://stackoverflow.com/a/42958644/1048572) [recommended before](https://stackoverflow.com/a/52767083/1048572), as it cleaner (no race conditions, less chance of forgetting the promise, simpler syntax) than the alternative.

Maybe introduce the closer-to-the-solution pattern only in the "Semantics as desugaring" section below?

Reply:
What do you think of the text I just uploaded, which includes your pattern?

Reply:
Thanks, perfect!